ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -431,11 +600,17 @@ S3Terminate(WT_STORAGE_SOURCE *storageSource, WT_SESSION *session)
         S3_FILE_SYSTEM *fs = s3->fsList.front();
         S3FileSystemTerminate(&fs->fileSystem, session);
     }
+    /*
+     * We should be single-threaded here. It should be safe to access the file handle list without a
+     * lock.
+     */
+    for (auto &fh : s3->fhList)
```

Comment:
Let's be consistent with the code that clears the lists, and we can update the comment for the file system list to include the handle list as well. You will have to close the file handles *before* closing the file systems. I was deliberately trying not to walk the list, since we are multi-threaded here - just that no thread refers to a storage source reference. A parallel thread could still close a file system / file handle and race with us.

I think we should re-do these two list walks under a lock. My hesitation comes from not knowing the tiered storage well enough to know what multi-threaded calls could come into the storage source and this point, when storage source is being terminated.

Reply:
If we end up taking a lock here, is it then ok to traverse the list? Agreed - it's unclear what will be happening in the threads at this point.